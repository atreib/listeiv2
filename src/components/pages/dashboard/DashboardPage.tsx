import React, { useState } from 'react';
import {
  ShoppingList,
  ListTitle,
  NewProduct,
  ListWrapper,
  Product,
  AppInput,
  AppButton,
  AppList,
} from './DashboardPage.styles';
import { ProductModel } from './../../../models';
import { generate } from './../../../helpers/uuid';
import { colors } from './../../../helpers/theme';

export const DashboardPage = () => {
  const [shoppingList, setShoppingList] = useState<ProductModel[]>();
  const [newProduct, setNewProduct] = useState<string>('');

  /**
   * Adds a new product to the shopping list state
   * * Uses the value typed on newProduct's state
   */
  const onAddNewProduct = () => {
    if (newProduct && newProduct !== '') {
      const newitem = {
        id: generate(),
        label: newProduct,
        quantity: 1,
        unityPrice: 0.0,
        totalPrice: 0.0,
      };

      if (shoppingList) setShoppingList([...shoppingList, newitem]);
      if (!shoppingList) setShoppingList([newitem]);
      setNewProduct('');
    }
  };

  /**
   * Removes a product on state
   * @param id: (number) id of the removed product
   */
  const removeProduct = (id: string) => {
    if (shoppingList) setShoppingList(shoppingList.filter((x) => x.id !== id));
  };

  /**
   * Change product quantity on state
   * @param increase: (boolean) if should increase or decrease product quantity
   * @param id: (string) id of the product being modified
   */
  const changeProductQuantity = (increase: boolean, id: string) => {
    if (shoppingList)
      setShoppingList(
        shoppingList.map((x) => {
          if (x.id === id) {
            return {
              ...x,
              quantity: increase ? x.quantity + 1 : x.quantity === 1 ? 1 : x.quantity - 1,
            };
          }
          return x;
        }),
      );
  };

  return (
    <ShoppingList>
      <ListTitle>Lista de compras</ListTitle>
      <ListWrapper>
        {shoppingList && (
          <AppList>
            {shoppingList.map((product) => (
              <Product
                changeProductQuantity={changeProductQuantity}
                onRemoveProduct={removeProduct}
                product={product}
                key={product.id}
              />
            ))}
          </AppList>
        )}
        {!shoppingList && 'Lista vazia'}
      </ListWrapper>
      <NewProduct>
        <AppInput
          data-testid="inputProduct"
          label="Novo produto"
          placeholder="Digite um produto"
          value={newProduct}
          setValue={setNewProduct}
        />
        <AppButton
          data-testid="btnAddProduct"
          bgColor={colors.primary}
          fontColor={colors.contrastPrimary}
          onClick={onAddNewProduct}
        >
          Add
        </AppButton>
      </NewProduct>
    </ShoppingList>
  );
};

export default DashboardPage;
