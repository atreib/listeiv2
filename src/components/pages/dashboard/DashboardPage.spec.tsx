import React from 'react';
import { shallow } from 'enzyme';
import { DashboardPage } from './DashboardPage';

describe('Dashboard page testes', () => {
  it('Should render correctly', () => {
    const component = shallow(<DashboardPage />);
    expect(component).toMatchSnapshot();
  });

  it('Should show page name on h1', () => {
    const component = shallow(<DashboardPage />);
    const z = component.find('h1').length;
    expect(z).toBe(1);
  });
});
