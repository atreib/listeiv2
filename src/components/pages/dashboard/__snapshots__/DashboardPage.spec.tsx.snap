// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard page testes Should render correctly 1`] = `
<h1>
  Dashboard
</h1>
`;
